In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_data = requests.get(url).text

Get Toronto postal code data from wikipedia with the url above

In [3]:
soup = BeautifulSoup(html_data,'html5lib')

Using the BeautifulSoup library to perform web scraping on the specified url page

In [4]:
raw = []
for row in soup.find('table').find_all('td'):
  raw_row = {}
  if row.find('span').text == 'Not assigned':
    pass

  else:    
    var = row.find('span').text.split('(')

    raw_row['Postal Code'] = row.find('b').text
    raw_row['Borough'] = var[0]
    raw_row['Neighborhood'] = var[1]

    raw.append(raw_row)

It can be seen that the data you want to get is in the table, in the table it can be seen that in the span group there are borough and neighborhood, while for the Postal Code it is in the first 'b' group.

In [5]:
df = pd.DataFrame(raw)
df.head()

Postal Code           Borough                        Neighborhood
0         M3A        North York                          Parkwoods)
1         M4A        North York                   Victoria Village)
2         M5A  Downtown Toronto         Regent Park / Harbourfront)
3         M6A        North York  Lawrence Manor / Lawrence Heights)
4         M7A      Queen's Park      Ontario Provincial Government)

In [6]:
df['Neighborhood'] = df['Neighborhood'].str.replace(r'\)$|\s$','').str.replace(r'\)',' ').str.replace(r'\s/',', ')
df['Borough'] = df['Borough'].replace({
    'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
    'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
    'EtobicokeNorthwest':'Etobicoke Northwest',
    'East YorkEast Toronto':'East York/East Toronto',
    'MississaugaCanada Post Gateway Processing Centre':'Mississauga'
    })

Neighborhood features are still very dirty with ')', '/', ''. Therefore, it is cleaned and tidied using replace and regex.
For Borough, there are some data that are not suitable, such as PO Boxes and others, so they need to be filtered and replaced with the appropriate data 

In [7]:
df

Postal Code  ...                                       Neighborhood
0           M3A  ...                                          Parkwoods
1           M4A  ...                                   Victoria Village
2           M5A  ...                         Regent Park,  Harbourfront
3           M6A  ...                  Lawrence Manor,  Lawrence Heights
4           M7A  ...                      Ontario Provincial Government
..          ...  ...                                                ...
98          M8X  ...    The Kingsway,  Montgomery Road,  Old Mill North
99          M4Y  ...                               Church and Wellesley
100         M7Y  ...                                     Enclave of M4L
101         M8Y  ...  Old Mill South,  King's Mill Park,  Sunnylea, ...
102         M8Z  ...  Mimico NW,  The Queensway West,  South of Bloo...

[103 rows x 3 columns]

In [8]:
df.shape

(103, 3)

---


In [18]:
!wget -q -O 'GeoSpatial_Dataset.csv' https://cocl.us/Geospatial_data
print('Dataset Downloaded')

Dataset Downloaded


Download the dataset regarding the latitude and longitude of the Toronto postal code. Because if you use the geocoder package it takes a very long time.

In [10]:
df_geo = pd.read_csv('GeoSpatial_Dataset.csv')

In [11]:
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [12]:
df_geo.shape

(103, 3)

It can be seen that the geospatial dataset has the same amount of data as the previous data, namely the postal code data in Toronto regarding the borough and neighborhood.

In [13]:
df_merge = df.merge(df_geo, how='inner', on='Postal Code')

The dataset is combined with a key value in the form of a Postal code. or if you are familiar with SQL then the term uses an inner join with a key in the form of a Postal Code

In [14]:
df_merge

Postal Code                Borough  ...   Latitude  Longitude
0           M3A             North York  ...  43.753259 -79.329656
1           M4A             North York  ...  43.725882 -79.315572
2           M5A       Downtown Toronto  ...  43.654260 -79.360636
3           M6A             North York  ...  43.718518 -79.464763
4           M7A           Queen's Park  ...  43.662301 -79.389494
..          ...                    ...  ...        ...        ...
98          M8X              Etobicoke  ...  43.653654 -79.506944
99          M4Y       Downtown Toronto  ...  43.665860 -79.383160
100         M7Y  East Toronto Business  ...  43.662744 -79.321558
101         M8Y              Etobicoke  ...  43.636258 -79.498509
102         M8Z              Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

In [15]:
df_merge.shape

(103, 5)

---

In [26]:
import numpy as np
import pandas as pd
import json 
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [27]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Get the coordinate Toronto then visualize using folium

In [28]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_merge['Latitude'], df_merge['Longitude'], df_merge['Borough'], df_merge['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Set the credentials to using Foursquare API further

In [29]:
CLIENT_ID = 'MFJPYDVFHRYQAMLUFAV3LNBMYVQTQ00UC2ROX3LSQQJCJ4HS' # your Foursquare ID
CLIENT_SECRET = 'DZ5RXHSQGRUFZYTAQ3H5NDQMCU3NINZQFLHGYRZT5ANCMTJP' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MFJPYDVFHRYQAMLUFAV3LNBMYVQTQ00UC2ROX3LSQQJCJ4HS
CLIENT_SECRET:DZ5RXHSQGRUFZYTAQ3H5NDQMCU3NINZQFLHGYRZT5ANCMTJP


Set function to get venues categories in Toronto

In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
toronto_venues = getNearbyVenues(names=df_merge['Neighborhood'],
                                   latitudes=df_merge['Latitude'],
                                   longitudes=df_merge['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park,  Harbourfront
Lawrence Manor,  Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern,  Rouge
Don Mills North
Parkview Hill,  Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park,  Princess Gardens,  Martin Grove,  Islington,  Cloverdale
Rouge Hill,  Port Union,  Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate,  Bloordale Gardens,  Old Burnhamthorpe,  Markland Wood
Guildwood,  Morningside,  West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor,  Wilson Heights,  Downsview North
Thorncliffe Park
Richmond,  Adelaide,  King
Dufferin,  Dovercourt Village
Scarborough Village
Fairview,  Henry Farm,  Oriole
Northwood Park,  York University
The Danforth  East
Harbourfront East,  Union Station,  Toronto Islands
Little Portugal,  Trinity
Kennedy Park,  Ionview,  East Birchmount Park
Bayview Village

In [32]:
print(toronto_venues.shape)
toronto_venues.head()

(2106, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

toronto_venues is data about venues in Toronoro for each Neighborhood and have 2106 row and 7 feature.

In [98]:
toronto_venues.groupby('Neighborhood').head().shape

(424, 7)

There are 424 venues were returned for each neighborhood

In [34]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 270 uniques categories.


one hot encoding the toronto venues

In [56]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

Add neighborhood column back to dataframe and place in first column

In [58]:

toronto_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
toronto_onehot.insert(0, 'Neighborhood', toronto_venues['Neighborhood'])

In [59]:
toronto_onehot.head()

Neighborhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  BBQ Joint  Baby Store  \
0                 0                   0            0          0           0   
1                 0                   0            0          0           0   
2                 0                   0            0          0           0   
3                 0                   0            0          0           0   
4                 0                   0            0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Court  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0            0                0         0             0              0   
1            0                0         0             0              0   
2            0                0         0             0              0   
3            0                0         0             0              0   
4            0                0         0             0              0 

Checking the size of dataframe

In [60]:
toronto_onehot.shape

(2106, 270)

Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [61]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                          Agincourt            0.000000   
1                             Alderwood,  Long Branch           0.000000   
2   Bathurst Manor,  Wilson Heights,  Downsview North           0.000000   
3                                     Bayview Village           0.000000   
4                  Bedford Park,  Lawrence Manor East           0.000000   
5                                         Berczy Park           0.000000   
6                        Birch Cliff,  Cliffside West           0.000000   
7      Brockton,  Parkdale Village,  Exhibition Place           0.000000   
8   CN Tower,  King and Spadina,  Railway Lands,  ...           0.000000   
9                                 Caledonia-Fairbanks           0.000000   
10                                          Cedarbrae           0.000000   
11                                 Central Bay Street           0.000000   
12                                           Christie           0.000000   
13                               Church and Wellesley           0.000000   
14  Clairville,  Humberwood,  Woodbine Downs,  Wes...           0.000000   
15          Clarks Corners,  Tam O'Shanter,  Sullivan           0.000000   
16  Cliffside,  Cliffcrest,  Scarborough Village West           0.000000   
17                    Commerce Court,  Victoria Hotel           0.000000   
18                                         Davisville           0.000000   
19                                   Davisville North           0.000000   
20  Del Ray,  Mount Dennis,  Keelsdale and Silvert...           0.000000   
21                                    Don Mills North           0.000000   
22                                    Don Mills South           0.000000   
23  Dorset Park,  Wexford Heights,  Scarborough To...           0.000000   
24                                  Downsview Central           0.000000   
25                                    Downsview East            0.000000   
26                                Downsview Northwest           0.000000   
27                                     Downsview West           0.000000   
28                      Dufferin,  Dovercourt Village           0.000000   
29                                     Enclave of L4W           0.000000   
30                                     Enclave of M4L           0.000000   
31                                     Enclave of M5E           0.000000   
32  Eringate,  Bloordale Gardens,  Old Burnhamthor...           0.000000   
33                     Fairview,  Henry Farm,  Oriole           0.000000   
34            First Canadian Place,  Underground city           0.000000   
35                           Forest Hill North & West           0.000000   
36                           Garden District, Ryerson           0.000000   
37                                          Glencairn           0.000000   
38                  Golden Mile,  Clairlea,  Oakridge           0.000000   
39                Guildwood,  Morningside,  West Hill           0.000000   
40  Harbourfront East,  Union Station,  Toronto Is...           0.000000   
41                     High Park,  The Junction South           0.000000   
42                                  Hillcrest Village           0.000000   
43                                      Humber Summit           0.000000   
44                                  Humberlea,  Emery           0.000000   
45                                 Humewood-Cedarvale           0.000000   
46                    India Bazaar,  The Beaches West           0.000000   
47      Kennedy Park,  Ionview,  East Birchmount Park           0.000000   
48        Kensington Market,  Chinatown,  Grange Park           0.000000   
49  Kingsview Village,  St. Phillips,  Martin Grov...           0.000000   
50                  Lawrence Manor,  Lawrence Heights           0.083333   
51                                      Lawrence Park           0.000000   
52           

Set function to sort the venues in descending order

In [62]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Make new dataframe and display the top 10 venues for each neighborhood

In [69]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                         Agincourt         Breakfast Spot   
1                            Alderwood,  Long Branch           Pizza Place   
2  Bathurst Manor,  Wilson Heights,  Downsview North           Coffee Shop   
3                                    Bayview Village   Japanese Restaurant   
4                 Bedford Park,  Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                Lounge          Skating Rink  Latin American Restaurant   
1        Sandwich Place           Coffee Shop                        Pub   
2                  Bank         Deli / Bodega             Sandwich Place   
3    Chinese Restaurant                  Bank                       Café   
4           Coffee Shop    Italian Restaurant           Greek Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Ethiopian Restaurant             Drugstore   Dumpling Restaurant   
1                   Pharmacy                   Gym          Gourmet Shop   
2  Middle Eastern Restaurant            Restaurant     Mobile Phone Shop   
3           Doner Restaurant            Donut Shop             Drugstore   
4                Pizza Place            Restaurant                   Pub   

         8th Most Common Venue        9th Most Common Venue  \
0  Eastern European Restaurant            Electronics Store   
1                 Dessert Shop                  Escape Room   
2                        Diner                  Supermarket   
3          Dumpling Restaurant  Eastern European Restaurant   
4             Sushi Restaurant            Indian Restaurant   

  10th Most Common Venue  
0            Escape Room  
1      Electronics Store  
2       Sushi Restaurant  
3            Yoga Studio  
4              Juice Bar

Then cluster the neighborhood

In [70]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 4, 0, 0, 0,
       2, 0, 0, 2, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0,
       0, 0, 0, 2, 0, 0, 0, 0, 0, 2], dtype=int32)

In [71]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [73]:
toronto_merged = df_merge

# merge toronto_grouped with df_merge to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                       Neighborhood   Latitude  \
0         M3A        North York                          Parkwoods  43.753259   
1         M4A        North York                   Victoria Village  43.725882   
2         M5A  Downtown Toronto         Regent Park,  Harbourfront  43.654260   
3         M6A        North York  Lawrence Manor,  Lawrence Heights  43.718518   
4         M7A      Queen's Park      Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             2.0                  Park       Food & Drink Shop   
1 -79.315572             0.0           Pizza Place             Coffee Shop   
2 -79.360636             0.0           Coffee Shop                  Bakery   
3 -79.464763             0.0        Clothing Store  Furniture / Home Store   
4 -79.389494             0.0           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0           Yoga Studio            Escape Room            Donut Shop   
1          Hockey Arena  Portuguese Restaurant               Dog Run   
2                  Café                   Park        Breakfast Spot   
3     Accessories Store     Miscellaneous Shop              Boutique   
4           Yoga Studio                   Bank              Beer Bar   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0              Drugstore   Dumpling Restaurant  Eastern European Restaurant   
1       Doner Restaurant            Donut Shop                    Drugstore   
2                Theater            Restaurant                          Pub   
3  Vietnamese Restaurant           Coffee Shop                   Donut Shop   
4          Smoothie Shop         Burrito Place               Sandwich Place   

  9th Most Common Venue       10th Most Common Venue  
0     Electronics Store                  Event Space  
1   Dumpling Restaurant  Eastern European Restaurant  
2           Yoga Studio                   Shoe Store  
3             Drugstore          Dumpling Restaurant  
4                  Café                   Restaurant

Checking if any Null value in dataframe

In [76]:
toronto_merged.isna().sum()

Postal Code               0
Borough                   0
Neighborhood              0
Latitude                  0
Longitude                 0
Cluster Labels            5
1st Most Common Venue     5
2nd Most Common Venue     5
3rd Most Common Venue     5
4th Most Common Venue     5
5th Most Common Venue     5
6th Most Common Venue     5
7th Most Common Venue     5
8th Most Common Venue     5
9th Most Common Venue     5
10th Most Common Venue    5
dtype: int64

Delete some rows that contain null value

In [77]:
toronto_merged.dropna(inplace=True)

Visualize the cluster

In [79]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Find out every cluster

### Cluster 1

In [81]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1                North York             0.0                 Pizza Place   
2          Downtown Toronto             0.0                 Coffee Shop   
3                North York             0.0              Clothing Store   
4              Queen's Park             0.0                 Coffee Shop   
6               Scarborough             0.0        Fast Food Restaurant   
7                North York             0.0  Construction & Landscaping   
8                 East York             0.0                 Pizza Place   
9          Downtown Toronto             0.0              Clothing Store   
10               North York             0.0                 Pizza Place   
13               North York             0.0                  Restaurant   
14                East York             0.0                Intersection   
15         Downtown Toronto             0.0                        Café   
16                     York             0.0                       Trail   
17                Etobicoke             0.0                 Pizza Place   
18              Scarborough             0.0                Intersection   
19             East Toronto             0.0           Health Food Store   
20         Downtown Toronto             0.0                 Coffee Shop   
22              Scarborough             0.0                 Coffee Shop   
23                East York             0.0                 Coffee Shop   
24         Downtown Toronto             0.0                 Coffee Shop   
25         Downtown Toronto             0.0               Grocery Store   
26              Scarborough             0.0                        Bank   
27               North York             0.0        Fast Food Restaurant   
28               North York             0.0                 Coffee Shop   
29                East York             0.0           Indian Restaurant   
30         Downtown Toronto             0.0                 Coffee Shop   
31             West Toronto             0.0                      Bakery   
32              Scarborough             0.0                  Playground   
33               North York             0.0              Clothing Store   
34               North York             0.0              Massage Studio   
36         Downtown Toronto             0.0                 Coffee Shop   
37             West Toronto             0.0                         Bar   
38              Scarborough             0.0          Chinese Restaurant   
39               North York             0.0         Japanese Restaurant   
40               North York             0.0  Construction & Landscaping   
41             East Toronto             0.0            Greek Restaurant   
42         Downtown Toronto             0.0                 Coffee Shop   
43             West Toronto             0.0                        Café   
44              Scarborough             0.0                    Bus Line   
46               North York             0.0                        Bank   
47             East Toronto             0.0        Fast Food Restaurant   
48         Downtown Toronto             0.0                 Coffee Shop   
49               North York             0.0  Construction & Landscaping   
50               North York             0.0                 Pizza Place   
51              Scarborough             0.0                       Motel   
53               North York             0.0           Korean Restaurant   
54             East Toronto             0.0                 Coffee Shop   
55               North York             0.0              Sandwich Place   
56                     York             0.0          Turkish Restaurant   
58              Scarborough             0.0                        Café   
59               North York             0.0                 Coffee Shop   
60               North York             0.0              Discount Store   
61          Central Toronto             0.0  

### Cluster 2

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto             1.0                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62     Electronics Store               Dog Run      Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62            Donut Shop             Drugstore   Dumpling Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62  Eastern European Restaurant           Yoga Studio    Filipino Restaurant

### Cluster 3

In [83]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             2.0                  Park   
21                    York             2.0                  Park   
35  East York/East Toronto             2.0     Convenience Store   
64                    York             2.0                  Park   
66              North York             2.0                  Park   
83         Central Toronto             2.0                  Park   
85             Scarborough             2.0                  Park   
91        Downtown Toronto             2.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio           Escape Room   
21         Women's Store                  Pool           Yoga Studio   
35                  Park           Yoga Studio           Escape Room   
64           Yoga Studio           Escape Room      Doner Restaurant   
66     Convenience Store           Yoga Studio           Escape Room   
83           Yoga Studio           Escape Room      Doner Restaurant   
85          Intersection            Playground     Electronics Store   
91                 Trail            Playground           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Donut Shop             Drugstore   Dumpling Restaurant   
21     Electronics Store      Doner Restaurant            Donut Shop   
35            Donut Shop             Drugstore   Dumpling Restaurant   
64            Donut Shop             Drugstore   Dumpling Restaurant   
66            Donut Shop             Drugstore   Dumpling Restaurant   
83            Donut Shop             Drugstore   Dumpling Restaurant   
85               Dog Run      Doner Restaurant            Donut Shop   
91               Dog Run      Doner Restaurant            Donut Shop   

          8th Most Common Venue 9th Most Common Venue  \
0   Eastern European Restaurant     Electronics Store   
21                    Drugstore   Dumpling Restaurant   
35  Eastern European Restaurant     Electronics Store   
64  Eastern European Restaurant     Electronics Store   
66  Eastern European Restaurant     Electronics Store   
83  Eastern European Restaurant     Electronics Store   
85                    Drugstore   Dumpling Restaurant   
91                    Drugstore   Dumpling Restaurant   

         10th Most Common Venue  
0                   Event Space  
21  Eastern European Restaurant  
35                  Event Space  
64         Ethiopian Restaurant  
66                  Event Space  
83         Ethiopian Restaurant  
85  Eastern European Restaurant  
91  Eastern European Restaurant

### Cluster 4

In [84]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12          Scarborough             3.0                   Bar   
94  Etobicoke Northwest             3.0         Garden Center   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12           Yoga Studio           Escape Room            Donut Shop   
94                   Bar   Rental Car Location             Drugstore   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
12             Drugstore   Dumpling Restaurant  Eastern European Restaurant   
94   Distribution Center      Doner Restaurant                   Donut Shop   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
12     Electronics Store         Ethiopian Restaurant      Fish & Chips Shop  
94   Dumpling Restaurant  Eastern European Restaurant      Electronics Store

### Cluster 5

In [85]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             4.0        Baseball Field          Food Service   
101   Etobicoke             4.0             Locksmith        Baseball Field   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57            Yoga Studio           Escape Room            Donut Shop   
101           Yoga Studio  Ethiopian Restaurant             Drugstore   

    6th Most Common Venue        7th Most Common Venue  \
57              Drugstore          Dumpling Restaurant   
101   Dumpling Restaurant  Eastern European Restaurant   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
57   Eastern European Restaurant     Electronics Store   Ethiopian Restaurant  
101            Electronics Store           Escape Room            Event Space